module4/lessons/Week2/BuildAnAPI.md

```diff
@@ -0,0 +1,403 @@
+* 🌶️ If the book does not exist, return a status code of 404
+</section>
+
+</section>
+
+### POST /books
+
+<section class='note' markdown='1'>
+
+**Implement an API Endpoint that satisfies these specifications**  
+* Path/Method: GET /api/books/1
+* Status Code: 200
+* Response Content: JSON containing the book with the id '1' (or whatever id is sent in through the path)
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### 1. Write a Test
+
+```c#
+[Fact]
+public async void PostBook_CreatesBookInDb()
+{
+    // Create fresh database
+    GoodBooksApiContext context = GetDbContext();
+
+    // Set up and send the request
+    HttpClient client = _factory.CreateClient();
+    var jsonString = "{\"Title\":\"Lamb's Wool\", \"Description\":\"Super Itchy\"}";
+    var requestContent = new StringContent(jsonString, Encoding.UTF8, "application/json");
+    var response = await client.PostAsync("/api/books", requestContent);
+
+    // Get the first (and should be only) book from the db
+    var newBook = context.Books.First();
+
+    Assert.Equal("Created", response.StatusCode.ToString());
+    Assert.Equal(201, (int)response.StatusCode);
+    Assert.Equal("Lamb's Wool", newBook.Title);
+}
+```
+
+A couple of things to note about this test!
+- We need to `GetDbContext` _before_ sending the post request, if we send it after, that method will delete the record we tried to create!
+- We can see two different ways to test the appropriate status code that is being returned.  This is because `response.StatusCode` returns and object which represents the status code and status phrase; `ToString()` returns the phrase associated with the code, and casting that object to an int reveals the integer code.
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### 2. Implement the Endpoint
+
+```c#
+[HttpPost]
+public ActionResult CreateBook(Book book)
+{
+    if (!ModelState.IsValid)
+    {
+        Response.StatusCode = 400;
+        return;
+    }
+    _context.Books.Add(book);
+    _context.SaveChanges();
+
+    var savedBook = _context.Books.Last();
+
+    Response.StatusCode = 201;
+    return new JsonResult(savedBook);
+}
+```
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### 3. Experiment with Postman
+
+Let's experiment with this new functionality in Postman.
+
+1. Open postman and create a post request
+2. Add a JSON body to the request
+3. Send the Request - you should receive a 201 response status!
+![](/assets/images/module4/week2/postman_postbooks.png)
+4. Verify the post by running a GetBooks request
+![](/assets/images/module4/week2/postman_getnewbooks.png)
+
+</section>
+
+### PUT /books/1
+
+<section class='call-to-action' markdown='1'>
+**With a partner: Implement an API Endpoint that satisfies these specifications**  
+
+<section class='note' markdown='1'>
+* Path/Method: PUT /api/books/1
+* Status Code: 204
+* Request Content: JSON with the ID of a book to edit, and the field(s) to be edited.
+</section>
```

Comment:
The formatting appears off in this section for some reason. The bullets do not render correctly.